src/picongpu/include/particles/ionization/byField/ADK/AlgorithmADKLinPol.hpp

```diff
@@ -0,0 +1,120 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include "simulation_defines.hpp"
+#include "particles/traits/GetAtomicNumbers.hpp"
+#include "particles/traits/GetIonizationEnergies.hpp"
+#include "traits/attribute/GetChargeState.hpp"
+#include "algorithms/math/floatMath/floatingPoint.tpp"
+#include "particles/ionization/utilities.hpp"
+
+/** \file AlgorithmADK.hpp
+ *
+ * IONIZATION ALGORITHM for the ADK model
+ *
+ * - implements the calculation of ionization probability and changes charge states
+ *   by decreasing the number of bound electrons
+ * - is called with the IONIZATION MODEL, specifically by setting the flag in @see speciesDefinition.param */
+
+namespace picongpu
+{
+namespace particles
+{
+namespace ionization
+{
+
+    /** Calculation for the Ammosov-Delone-Krainov tunneling model
+     *
+     * for linear laser polarization
+     */
+    struct AlgorithmADKLinPol
+    {
+
+        /** Functor implementation
+         * \tparam EType type of electric field
+         * \tparam BType type of magnetic field
+         * \tparam ParticleType type of particle to be ionized
+         *
+         * \param bField magnetic field value at t=0
+         * \param eField electric field value at t=0
+         * \param parentIon particle instance to be ionized with position at t=0 and momentum at t=-1/2
+         */
+        template<typename EType, typename BType, typename ParticleType >
+        HDINLINE void
+        operator()( const BType bField, const EType eField, ParticleType& parentIon, float_X randNr )
+        {
+
+            const float_X protonNumber = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            float_X chargeState = attribute::getChargeState(parentIon);
+
+            /* verify that ion is not completely ionized */
+            if (chargeState < protonNumber)
+            {
+                uint32_t cs = math::float2int_rd(chargeState);
+                const float_X iEnergy = GetIonizationEnergies<ParticleType>::type()[cs];
+
+                const float_X pi = precisionCast<float_X>(M_PI);
+                /* electric field in atomic units - only absolute value */
+                float_X eInAU = math::abs(eField) / ATOMIC_UNIT_EFIELD;
+
+                /* effective principal quantum number (unitless) */
+                float_X nEff = protonNumber / math::sqrt(float_X(2.0) * iEnergy );
+                /* nameless variable for convenience dFromADK*/
+                float_X dBase = float_X(4.0) * util::cube(protonNumber) / (eInAU * util::quad(nEff)) ;
+                float_X dFromADK = math::pow(dBase,nEff);
+
+                /* factor from averaging over one laser cycle with LINEAR polarization */
+                float_X polarizationFactor = math::sqrt(float_X(3.0) * util::cube(nEff) * eInAU / (pi * util::cube(protonNumber)));
```

Comment:
since this line is the **only** difference between `AlgorithmADKCircPol`and `AlgorithmADKLinPol`, but the  code for `AlgorithmADKLinPol` takes about 120 additional lines, could you create a `AlgorithmADKPol<...>`that takes a template argument to switch between both cases? That would reduce the code duplication dramatically.